src/reconcile/rec_visibility.c

```diff
@@ -782,27 +771,145 @@ __rec_upd_select(WT_SESSION_IMPL *session, WTI_RECONCILE *r, WT_UPDATE *first_up
+                    *write_prepare = true;
+                }
+            } else if (upd->upd_durable_ts > r->rec_start_pinned_stable_ts) {
+                WT_ASSERT(session, !is_hs_page);
+                *upd_memsizep += WT_UPDATE_MEMSIZE(upd);
+                *has_newer_updatesp = true;
+                continue;
+            }
+        } else if (!F_ISSET(upd, WT_UPDATE_RESTORED_FROM_DELTA)) {
+            if (prepare_state == WT_PREPARE_INPROGRESS) {
+                WT_ASSERT_ALWAYS(session,
+                  upd_select->upd == NULL || upd_select->upd->txnid == upd->txnid,
+                  "Cannot have two different prepared transactions active on the same key");
+
+                if (F_ISSET(r, WT_REC_CHECKPOINT)) {
+                    *upd_memsizep += WT_UPDATE_MEMSIZE(upd);
+                    *has_newer_updatesp = true;
+                    seen_prepare = true;
+                    continue;
+                } else {
+                    *write_prepare = true;
+
+                    /*
+                     * If we are not in eviction, we must be in salvage to reach here. Since salvage
+                     * only works on data on-disk, the prepared update must be restored from the
+                     * disk. No need for us to rollback the prepared update in salvage here. If
+                     * there is still content for that key left in the history store, rollback to
+                     * stable will bring it back to the data store later. Otherwise, it removes the
+                     * key.
+                     */
+                    WT_ASSERT_ALWAYS(session,
+                      F_ISSET(r, WT_REC_EVICT) ||
+                        (F_ISSET(r, WT_REC_VISIBILITY_ERR) &&
+                          F_ISSET(upd, WT_UPDATE_PREPARE_RESTORED_FROM_DS)),
+                      "Should never salvage a prepared update not from disk.");
+                    /* Prepared updates cannot be resolved concurrently to eviction and salvage. */
+                    WT_ASSERT_ALWAYS(session, upd->prepare_state == WT_PREPARE_INPROGRESS,
+                      "Should never concurrently resolve a prepared update during reconciliation "
+                      "if we "
+                      "are not in a checkpoint.");
+                }
+            }
         }
 
+        if (F_ISSET(conn, WT_CONN_PRESERVE_PREPARED) && F_ISSET(upd, WT_UPDATE_PREPARE_ROLLBACK) &&
+          !F_ISSET(upd, WT_UPDATE_SELECT_FOR_DS))
+            prepare_rollback_tombstone = upd;
+
         /*
          * Always select the newest visible update if precise checkpoint is not enabled. Otherwise,
          * select the first update that is smaller or equal to the pinned timestamp.
          */
         if (upd_select->upd == NULL)
             upd_select->upd = upd;
+        else if (prepare_rollback_tombstone != NULL) {
+            WT_ASSERT(session,
+              *write_prepare && upd->txnid == WT_TXN_ABORTED &&
```

Comment:
When doing prepared rollback, we first insert the rollback tombstone then mark the prepared as aborted. So this assert can fire if we race with prepared rollback. Remove the check on the txnid and add a comment explain why.

Reply:
done